lib/engine.js

```diff
@@ -392,13 +392,19 @@ class GameEngine {
 
   invalidateCache() {
     this.lastValidTime = 0;
-    this.statesByTime = {};
+    this.statesByTime = new Map();
   }
 
   addEvent(eventTime, name, ...args) {
     const events = this.eventsByTime[eventTime] || [];
 
-    this.lastValidTime = Math.min(this.lastValidTime, eventTime);
+    if (eventTime <= this.lastValidTime - STATE_CACHE_SIZE) {
+      console.log("Event underrun detected");
```

Comment:
You shouldn't log anything into console unless some kind of debugging mode is on.